pintc/src/predicate/transform/dead_code_elimination.rs

```diff
@@ -0,0 +1,39 @@
+use crate::{
+    expr::Expr,
+    predicate::{Contract, StateKey},
+};
+
+/// In a given contract, remove any code that is not reachable or used.
+pub(crate) fn dead_code_elimination(contract: &mut Contract) {
+    dead_state_elimination(contract);
+}
+
+/// Remove all unused States in their respective predicates.
+pub(crate) fn dead_state_elimination(contract: &mut Contract) {
+    for pred_key in contract.preds.keys().collect::<Vec<_>>() {
+        let mut live_paths: Vec<String> = Vec::new();
+        for expr in contract.exprs(pred_key) {
+            if let Expr::Path(name, _) = expr.get(contract) {
+                live_paths.push(name.to_string());
+            }
+        }
```

Comment:
If you want to practice some functional-style Rust, you can try to rewrite this using a `filter_map` and then `collect` into a `HashSet`.